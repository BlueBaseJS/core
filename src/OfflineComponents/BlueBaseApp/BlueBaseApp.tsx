import { BlueBase, BlueBaseProgress, BootOptions } from '../../BlueBase';
import {
	BlueBaseAppError,
	BlueBaseAppErrorProps,
	BlueBaseAppLoading,
	BlueBaseAppLoadingProps,
} from '..';
import React, { useEffect, useState } from 'react';

import { BlueBaseProvider } from '../../Context';
import { IntlProvider } from '../../intl';
import { ThemeProvider } from '../../themes';
import { getComponent } from '../../getComponent';

const BlueBaseContent = getComponent('BlueBaseContent');

export interface BlueBaseAppProps extends Partial<BootOptions> {
	/** BlueBase Context */
	BB?: BlueBase;

	/** Component rendered when an error occurs */
	ErrorComponent: React.ComponentType<BlueBaseAppErrorProps>;

	/** Component rendered when an error occurs */
	LoadingComponent: React.ComponentType<BlueBaseAppLoadingProps>;

	/**
	 * If this prop is provided, this tree will be rendered instead of BlueBase's own view.
	 */
	children?: React.ReactNode;

	/**
	 * Used to locate this view in end-to-end tests.
	 */
	testID?: string;
}

/**
 * # ðŸš€ BlueBaseApp
 *
 * The main BlueBase app. This is the top level component in BlueBase. Takes care
 * of initialisation, and renders either children, or app with routing.
 *
 * ## Usage
 * ```jsx
 * <BlueBaseApp BB={BB} />
 * ```
 */
export const BlueBaseApp = (props: BlueBaseAppProps) => {
	const {
		ErrorComponent = BlueBaseAppError,
		LoadingComponent = BlueBaseAppLoading,
		children,
	} = props;

	const [bootCount, setBootCount] = useState(0);
	const [booting, setBooting] = useState(true);
	const [bootTrigger, setBootTrigger] = useState(true); // Setting to true to start boot
	const [progress, setProgress] = useState<BlueBaseProgress>({});

	const [BB] = useState(props.BB || new BlueBase());

	BB.reboot = async () => {
		setBootTrigger(true);
	};

	useEffect(() => {
		if (bootTrigger) {
			setBootTrigger(false);
		} else {
			return;
		}

		(async () => {
			await BB.boot({
				...props,
				onProgress: (p: BlueBaseProgress) => {
					setProgress(p);
					if (p.error) {
						setBooting(false);
					}
				},
			});
			setBootCount(bootCount + 1);
			setBooting(false);
		})();
	}, [bootTrigger]);

	if (booting) {
		return <LoadingComponent BB={BB} progress={progress} bootCount={bootCount} />;
	}

	if (progress.error) {
		return <ErrorComponent BB={BB} progress={progress} bootCount={bootCount} />;
	}

	return (
		<BlueBaseProvider key={`boot-${bootCount}`} value={BB}>
			<ThemeProvider>
				<IntlProvider>
					<BlueBaseContent>{children}</BlueBaseContent>
				</IntlProvider>
			</ThemeProvider>
		</BlueBaseProvider>
	);
};

const defaultProps: Partial<BlueBaseAppProps> = {
	ErrorComponent: BlueBaseAppError,
	LoadingComponent: BlueBaseAppLoading,
};

BlueBaseApp.defaultProps = defaultProps;
