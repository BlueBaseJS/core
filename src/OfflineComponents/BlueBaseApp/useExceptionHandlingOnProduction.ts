import { BlueBase } from '../../BlueBase';
import { isProduction } from '../../utils';

// Number of errors to save
const CACHE_SIZE = 5;

// If a successive error take longer than this time, we don't consider it as recursive
const RECURSIVE_PERIOD = 300;

let errors: Array<{
	error: any;
	time: number;
}> = [];

/**
 * If the app is in production mode, automatically reboots BB on exception.
 * Shows an error message if the app is stuck in infinite crash.
 * @param BB
 */
export function useExceptionHandlingOnProduction(BB: BlueBase) {
	let development = BB.Configs.getValue('development');

	if (development === undefined) {
		development = !isProduction();
	}

	/**
	 * Adds an error to the errors array, with the timestamp
	 * @param e
	 */
	function addError(e: any) {
		const arr = [
			...errors,
			{
				error: e,
				time: new Date().getTime(),
			},
		];

		errors = arr.length > CACHE_SIZE ? arr.slice(1) : arr;
	}

	/**
	 * Gets periods of time in successive errors
	 */
	function isInfinite() {
		if (errors.length < CACHE_SIZE) {
			return false;
		}

		for (let i = 1; i < errors.length; i++) {
			if (errors[i].time - errors[i - 1].time > RECURSIVE_PERIOD) {
				return false;
			}
		}

		return true;
	}

	function onError(e: any) {
		addError(e);

		if (!development && !isInfinite()) {
			BB.reboot();
		}
	}

	return { onError };
}
