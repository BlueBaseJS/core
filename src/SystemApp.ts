import { BlueRain, JsonComponentSchema, withBlueRain } from './index';
import React from 'react';

export const SystemRoutes = (BR: BlueRain) => {
	const appRoutes = BR.Apps.getAllRoutes();

	// Default Routes
	let systemRoutes: JsonComponentSchema[] = [
		{
			component: 'Route',
			props: {
				path: '/',
				exact: true,
				component: BR.Components.get('IndexPage')
			}
		},
		...appRoutes
	];

	// Dynamically add new routes
	systemRoutes = BR.Filters.run('bluerain.system.routes', systemRoutes);

	// Add 404 at the end
	systemRoutes.push({
		component: 'Route',
		props: {
			component: BR.Components.get('NotFoundPage')
		}
	});

	return systemRoutes;
};

const SystemApp = (props: { bluerain: BlueRain }) => {
	const BR = props.bluerain;

	let routes: JsonComponentSchema = {
		component: 'SystemLayout',
		props,
		children: [
			{
				component: 'RouterSwitch',
				props: { style: { flex: 1, flexGrow: 1 } },
				children: SystemRoutes(BR)
			}
		]
	};

	routes = BR.Filters.run('bluerain.system.app.schema', routes);
	return BR.API.JsonToReact.parse(routes);
};

export default withBlueRain(SystemApp);
