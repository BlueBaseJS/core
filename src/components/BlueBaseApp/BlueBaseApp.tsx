import { BlueBase, BootOptions } from '../../BlueBase';
import React from 'react';
import { Text } from 'react-native';

export interface BlueBaseAppProps extends Partial<BootOptions> {
	BB?: BlueBase,
	children?: React.ReactNode,
}

interface BlueBaseAppState {

	booted: boolean;
	BB: BlueBase

	Component: React.ComponentType<any>
}

/**
 * ðŸš€ BlueBaseApp
 *
 * This is the Main App used to render at the top level.
 *
 * TODO: Add better docs
 */
// NOTE FOR DEVELOPERS:
// Don't use BlueBase context or any data saved in context (e.g. components)
// here. This is because the context may not be initialized yet.
export class BlueBaseApp extends React.Component<BlueBaseAppProps, BlueBaseAppState> {

	constructor(props: BlueBaseAppProps) {
		super(props);

		this.state = {
			BB: props.BB || new BlueBase(),
			Component: () => <Text>Loading</Text>,
			booted: false,
		};
	}

	async componentDidMount() {

		const BB = this.state.BB;

		try {
			const Component = await BB.boot(this.props);
			this.setState({
				Component,
				booted: BB.booted,
			});
		} catch (error) {
			// tslint:disable-next-line:no-console
			console.error(error);
		}
	}

	render() {

		const Component = this.state.Component;

		return <Component children={this.props.children} />;
	}
}