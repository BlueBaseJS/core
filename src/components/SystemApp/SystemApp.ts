import { BlueRain, JsonComponentSchema, withBlueRain } from '../../index';
import React from 'react';

/**
 * Shows system content
 */
const SystemApp = ({ bluerain: BR, ...others }: { bluerain: BlueRain }) => {
	let routes: JsonComponentSchema = {
		component: 'SystemLayout',
		props: others,
		children: [
			{
				component: 'SystemHeader'
			},
			{
				component: 'SystemContent'
			},
			{
				component: 'SystemFooter'
			}
		]
	};

	routes = BR.Filters.run('bluerain.system.app.schema', routes);
	return BR.API.JsonToReact.parse(routes);
};

export default withBlueRain(SystemApp) as React.ComponentType<any>;
