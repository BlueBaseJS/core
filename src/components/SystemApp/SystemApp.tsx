import { View, ViewProperties, ViewStyle } from 'react-native';
import { BlueBase } from '../../BlueBase';
import { BlueBaseContext } from '../../Context';
import React from 'react';

export interface SystemAppProps extends ViewProperties {
	styles: {
		content: ViewStyle,
		footer: ViewStyle,
		header: ViewStyle,
		root: ViewStyle,
	}
}

/**
 * Main SystemApp. This is the component that renders content to the screen.
 */
export class SystemApp extends React.PureComponent<SystemAppProps> {

	static contextType = BlueBaseContext;

	static defaultStyles = {

		// SystemContent styles
		content: {
			flex: 1,
		},

		// SystemFooter styles
		footer: {},

		// SystemHeader styles
		header: {},

		// Root container styles
		root: {
			flex: 1,
		},
	};

	render() {

		const BB: BlueBase = this.context;

		const { children, style, styles, ...rest } = this.props;

		const SystemHeader = BB.Components.resolve('SystemHeader');
		const SystemContent = BB.Components.resolve('SystemContent');
		const SystemFooter = BB.Components.resolve('SystemFooter');

		return (
			<View style={[styles.root, style]} {...rest}>
				<SystemHeader style={styles.header} />
				<SystemContent style={styles.content}>{children}</SystemContent>
				<SystemFooter style={styles.footer} />
			</View>
		);
	}
}