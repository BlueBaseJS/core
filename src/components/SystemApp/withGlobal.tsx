import React from 'react';

const withGlobal = (WrappedCompopnent: React.ComponentType): React.ComponentType => {
	const _global: any = global;
	if (typeof _global.self === 'undefined') {
		_global.self = global;
	}
	return WrappedCompopnent;
};

export default withGlobal;
