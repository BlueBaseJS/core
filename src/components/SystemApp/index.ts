import SystemApp from './SystemApp';
export default SystemApp;
