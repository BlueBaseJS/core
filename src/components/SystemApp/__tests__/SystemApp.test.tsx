import { BlueBaseApp } from '../../BlueBaseApp';
import React from 'react';
import { SystemApp } from '../SystemApp';
import TestRenderer from 'react-test-renderer';

beforeEach(() => {
	jest.resetModules();
});

describe('SystemApp', () => {
	const SystemAppWithProvider = (props: any) => (
		<BlueBaseApp>
			<SystemApp {...props} />
		</BlueBaseApp>
	);

	test(`Snapshot SystemApp component`, () => {
		const component = TestRenderer.create(
			<SystemAppWithProvider />
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	test(`Snapshot SystemApp component after complete render`, (done) => {
		const component = TestRenderer.create(
			<SystemAppWithProvider />
		);
		setTimeout(() => {
			const tree = component.toJSON();
			expect(tree).toMatchSnapshot();
			done();
		});
	});
});
