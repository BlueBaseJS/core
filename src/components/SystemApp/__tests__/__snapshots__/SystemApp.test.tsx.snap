// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SystemApp Snapshot SystemApp component 1`] = `
<Text>
  Loading
</Text>
`;

exports[`SystemApp Snapshot SystemApp component after complete render 1`] = `
<View>
  <View>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontFamily": "Helvetica",
            "fontSize": 16,
            "fontWeight": "400",
          },
          undefined,
        ]
      }
    >
      ðŸš€ BlueBase System Content!
    </Text>
  </View>
</View>
`;
